// tsparticles/dist/Core/Retina.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Retina = void 0;
var Retina = (function () {
    function Retina(container) {
        this.container = container;
        this.bubbleModeDistance = 0;
        this.bubbleModeSize = 0;
        this.connectModeDistance = 0;
        this.connectModeRadius = 0;
        this.grabModeDistance = 0;
        this.repulseModeDistance = 0;
        this.slowModeRadius = 0;
        this.linksDistance = 0;
        this.linksWidth = 0;
        this.moveSpeed = 0;
        this.sizeValue = 0;
        this.sizeAnimationSpeed = 0;
        this.pixelRatio = 1;
    }
    Retina.prototype.init = function () {
        var _a;
        var container = this.container;
        var options = container.options;
        if (options.detectRetina) {
            this.pixelRatio = window.devicePixelRatio;
        }
        else {
            this.pixelRatio = 1;
        }
        var ratio = this.pixelRatio;
        if (container.canvas.element) {
            var element = container.canvas.element;
            container.canvas.size.width = element.offsetWidth * ratio;
            container.canvas.size.height = element.offsetHeight * ratio;
        }
        var particles = options.particles;
        this.linksDistance = particles.links.distance * ratio;
        this.linksWidth = particles.links.width * ratio;
        this.moveSpeed = particles.move.speed * ratio;
        this.sizeValue = particles.size.value * ratio;
        this.sizeAnimationSpeed = particles.size.animation.speed * ratio;
        var modes = options.interactivity.modes;
        this.connectModeDistance = modes.connect.distance * ratio;
        this.connectModeRadius = modes.connect.radius * ratio;
        this.grabModeDistance = modes.grab.distance * ratio;
        this.repulseModeDistance = modes.repulse.distance * ratio;
        this.slowModeRadius = modes.slow.radius * ratio;
        this.bubbleModeDistance = modes.bubble.distance * ratio;
        this.bubbleModeSize = ((_a = modes.bubble.size) !== null && _a !== void 0 ? _a : this.sizeValue * 2) * ratio;
    };
    Retina.prototype.initParticle = function (particle) {
        var particlesOptions = particle.particlesOptions;
        var ratio = this.pixelRatio;
        particle.linksDistance = particlesOptions.links.distance * ratio;
        particle.linksWidth = particlesOptions.links.width * ratio;
        particle.moveSpeed = particlesOptions.move.speed * ratio;
        particle.sizeValue = particlesOptions.size.value * ratio;
        if (typeof particlesOptions.size.random !== "boolean") {
            particle.randomMinimumSize = particlesOptions.size.random.minimumValue * ratio;
        }
        particle.sizeAnimationSpeed = particlesOptions.size.animation.speed * ratio;
    };
    Retina.prototype.reset = function () {
    };
    return Retina;
}());
exports.Retina = Retina;
